src/shared/rm-rf.c

```diff
@@ -115,6 +115,57 @@ int fstatat_harder(int dfd,
         return 0;
 }
 
+static int openat_harder(int dfd, const char *path, int open_flags, RemoveFlags remove_flags, mode_t *ret_old_mode) {
+        _cleanup_close_ int pfd = -EBADF, fd = -EBADF;
+        mode_t old_mode;
+        int r;
+
+        assert(dfd >= 0 || dfd == AT_FDCWD);
+        assert(path);
+        assert(open_flags & O_DIRECTORY);
+
+        /* Unlike unlink_harder() and fstatat_harder(), this chmod the specified path. */
+
+        fd = RET_NERRNO(openat(dfd, path, open_flags));
+        if (fd >= 0) {
+                if (ret_old_mode) {
+                        struct stat st;
+
+                        if (fstat(fd, &st) < 0)
+                                return -errno;
+
+                        *ret_old_mode = st.st_mode;
+                }
+
+                return TAKE_FD(fd);
+        }
+        if (fd != -EACCES || !FLAGS_SET(remove_flags, REMOVE_CHMOD))
+                return fd;
+
+        pfd = RET_NERRNO(openat(dfd, path, O_PATH|O_CLOEXEC|O_DIRECTORY|(open_flags & O_NOFOLLOW)));
+        if (pfd < 0)
+                return pfd;
+
```

Comment:
one could probably argue that it would be safer to always open with O_PATH first, and then use fd_reopen() from there. because otherwise the dentry might be swapped out between the first openat() and the second()...